### PR TITLE
ggx_smithcorrelated from frostbite 3.

### PR DESCRIPTION
**Pull request by bhouston**
Quick implementation of proposed feature here: https://github.com/mrdoob/three.js/issues/8355

I did simplify the formulate a bit, which is still mathematically equivalent, so that it resembles more closely our existing G_GGX_Smith function.

### CODE REVIEW

File: src/renderers/shaders/ShaderChunk/bsdfs.glsl [hunk 1]
```diff
@@ -60,13 +60,24 @@ float G_GGX_Smith( const in float alpha, const in float dotNL, const in float do
...
 
 } // validated
 
+// from page 12, listing 2 of http://www.frostbite.com/wp-content/uploads/2014/11/course_notes_moving_frostbite_to_pbr_v2.pdf
+float G_GGX_SmithCorrelated( const in float alpha, const in float dotNL, const in float dotNV ) {
+
+	float a2 = pow2( alpha );
+
+	float gv = dotNL * sqrt( a2 + ( 1.0 - a2 ) * pow2( dotNV ) );
+	float gl = dotNV * sqrt( a2 + ( 1.0 - a2 ) * pow2( dotNL ) );
```
Comment (WestLangley) [inline src/renderers/shaders/ShaderChunk/bsdfs.glsl]: maybe the comment would be helpful. :-)

    // dotNL and dotNV are explicitly swapped. This is not a mistake.

**Comment by WestLangley** [comment]
This version does appear to be very sensitive to `dotNL` and `dotNV` being zero -- regardless of the value of `roughness`.

**Comment by bhouston** [comment]
> This version does appear to be very sensitive to dotNL and dotNV being zero -- regardless of the value of roughness.

You are right.  In this case it is probably easier to add EPSILON to the denominator (or do a max( ...., EPSILON) rather than clamping the dotNV, dotNL, etc...

**Comment by bhouston** [comment]
I'd added the comment you suggested @WestLangley and also added a max(..., EPSILON) to the denominator.

**Comment by WestLangley** [comment]
Thanks, @bhouston.

**Comment by mrdoob** [comment]
Hmm, any reason to keep the old version?

**Comment by WestLangley** [comment]
>Hmm, any reason to keep the old version?

I think we should keep the old version around for comparison for awhile -- until @bhouston is convinced there are no issues with the new version.

**Comment by bhouston** [comment]
We could start making a set of *.glsl files in an "examples/glsl" directory of stuff that could theroetically be useful.  The more advanced cube_uv stuff comes to mind.

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**